platform/osx/os_osx.mm

```diff
@@ -591,18 +595,75 @@ static void _mouseDownEvent(NSEvent *event, int index, int mask, bool pressed) {
 - (void)mouseDown:(NSEvent *)event {
 	if (([event modifierFlags] & NSEventModifierFlagControl)) {
 		mouse_down_control = true;
+
 		_mouseDownEvent(event, BUTTON_RIGHT, BUTTON_MASK_RIGHT, true);
 	} else {
 		mouse_down_control = false;
-		_mouseDownEvent(event, BUTTON_LEFT, BUTTON_MASK_LEFT, true);
+		if (OS_OSX::singleton->drag_mode != OS::DRAG_MODE_MOVE) {
+			_mouseDownEvent(event, BUTTON_LEFT, BUTTON_MASK_LEFT, true);
+		}
 	}
 }
 
 - (void)mouseDragged:(NSEvent *)event {
-	[self mouseMoved:event];
+	// Prevent from being both movable and recieving drag events.
+	if ([[event window] isMovableByWindowBackground]) {
+		return;
+	}
+
+	NSRect frame = [[event window] frame];
+	switch (OS_OSX::singleton->drag_mode) {
+		case OS::DRAG_MODE_RESIZE_TOP:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x, frame.origin.y, frame.size.width, frame.size.height - event.deltaY)
+							 display:YES];
+			break;
+		case OS::DRAG_MODE_RESIZE_BOTTOM:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x, frame.origin.y - event.deltaY, frame.size.width, frame.size.height + event.deltaY)
+							 display:YES];
+			break;
+		case OS::DRAG_MODE_RESIZE_LEFT:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x + event.deltaX, frame.origin.y, frame.size.width - event.deltaX, frame.size.height)
+							 display:YES];
+			break;
+		case OS::DRAG_MODE_RESIZE_RIGHT:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x, frame.origin.y, frame.size.width + event.deltaX, frame.size.height)
+							 display:YES];
+			break;
+		case OS::DRAG_MODE_RESIZE_TOPLEFT:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x + event.deltaX, frame.origin.y, frame.size.width - event.deltaX, frame.size.height - event.deltaY)
+							 display:YES];
+			break;
+		case OS::DRAG_MODE_RESIZE_BOTTOMLEFT:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x + event.deltaX, frame.origin.y - event.deltaY, frame.size.width - event.deltaX, frame.size.height + event.deltaY)
+							 display:YES];
+			break;
+		case OS::DRAG_MODE_RESIZE_TOPRIGHT:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x, frame.origin.y, frame.size.width + event.deltaX, frame.size.height - event.deltaY)
+							 display:YES];
+			break;
+		case OS::DRAG_MODE_RESIZE_BOTTOMRIGHT:
+			[[event window] setFrame:NSMakeRect(
+											 frame.origin.x, frame.origin.y - event.deltaY, frame.size.width + event.deltaX, frame.size.height + event.deltaY)
+							 display:YES];
+			break;
+		default:
+			[self mouseMoved:event];
+	}
 }
 
 - (void)mouseUp:(NSEvent *)event {
+	if (OS_OSX::singleton->drag_mode == OS::DRAG_MODE_MOVE && [event clickCount] == 2) {
+		OS_OSX::singleton->set_window_maximized(!OS_OSX::singleton->is_window_maximized());
```

Comment:
On macOS double-clicking window title should minimize window not maximize, and there's nothing like this implemented on other platforms.